File: src/main/java/co/onmind/microhex/application/mappers/RoleMapper.java
package co.onmind.microhex.application.mappers;

import co.onmind.microhex.application.dto.CreateRoleRequest;
import co.onmind.microhex.application.dto.RoleResponse;
import co.onmind.microhex.domain.models.Role;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.factory.Mappers;

import java.util.List;

/**
 * MapStruct mapper for converting between Role domain models and DTOs.
 * 
 * This mapper uses MapStruct to automatically generate the implementation
 * at compile time, providing better performance and type safety.
 * 
 * @author OnMind (Cesar Andres Arcila Buitrago)
 * @version 2.0.0
 */
@Mapper(componentModel = "spring")
public interface RoleMapper {
    
    RoleMapper INSTANCE = Mappers.getMapper(RoleMapper.class);
    
    /**
     * Converts a Role domain model to RoleResponse DTO.
     * @param role the domain model
     * @return the response DTO
     */
    RoleResponse toResponse(Role role);
    
    /**
     * Converts a list of Role domain models to a list of RoleResponse DTOs.
     * @param roles the list of domain models
     * @return the list of response DTOs
     */
    List<RoleResponse> toResponseList(List<Role> roles);
    
    /**
     * Converts a CreateRoleRequest DTO to Role domain model.
     * Maps only the name field, ignoring id and createdAt.
     * @param request the create request DTO
     * @return the domain model
     */
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    @Mapping(target = "withName", ignore = true)
    Role toDomain(CreateRoleRequest request);
}

File: src/main/kotlin/co/onmind/hex/application/mapper/RoleMapper.kt
package co.onmind.hex.application.mapper

import co.onmind.hex.application.dto.RoleResponse
import co.onmind.hex.domain.model.Role
import jakarta.inject.Singleton

/**
 * Mapper for converting between domain models and DTOs.
 * 
 * This mapper handles the transformation between the domain layer
 * and the application layer, ensuring proper data conversion
 * while maintaining separation of concerns.
 */
@Singleton
class RoleMapper {
    
    /**
     * Converts a Role domain model to RoleResponse DTO.
     */
    fun toResponse(role: Role): RoleResponse {
        return RoleResponse(
            id = role.id ?: throw IllegalStateException("Role ID cannot be null for response"),
            name = role.name,
            createdAt = role.createdAt,
            isSystemRole = role.isSystemRole()
        )
    }
    
    /**
     * Converts a list of Role domain models to RoleResponse DTOs.
     */
    fun toResponseList(roles: List<Role>): List<RoleResponse> {
        return roles.map { toResponse(it) }
    }
}